module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+			fe->fe_logical_len = extent_end - end;
+			fe->fe_logical_start = end;
+		/*
+		 * Extent spans before and after the region to be clearer.
+		 * Split the extent in to a before and after portion.
+		 */
+		} else if (extent_start < start && extent_end > end) {
+			fe->fe_logical_len = start - extent_start;
+			zfs_fiemap_add_impl(t, end, extent_end - end,
+			    0, 0, fe->fe_vdev, fe->fe_flags);
+		} else {
+			fe = AVL_NEXT(t, fe);
+			continue;
+		}
+
+		/*
+		 * Zero the physical start and length which are no longer
+		 * meaningful after modifying the logical start or length.
+		 *
+		 * N.B. Ideally we should keep a list the block pointers
+		 * comprising the extent.  This would allow us to properly
+		 * trim it and correctly update the physical start and length.
+		 */
+		fe->fe_physical_start = 0;
+		fe->fe_physical_len = 0;
+
+		fe = AVL_NEXT(t, fe);
+	}
+}
+
+/*
+ * Pending dirty extents set FIEMAP_EXTENT_DELALLOC to indicate they have
+ * not yet been written.  The FIEMAP_EXTENT_UNKNOWN flag must be set when
+ * FIEMAP_EXTENT_DELALLOC is set.  Dirty extents are only inserted in to
+ * the first extent tree.
+ */
+static void
+zfs_fiemap_add_dirty(void *arg, uint64_t start, uint64_t size)
+{
+	zfs_fiemap_t *fm = (zfs_fiemap_t *)arg;
+	avl_tree_t *t = &fm->fm_extent_trees[0];
+
+	zfs_fiemap_clear(t, start, size);
+
+	if (fm->fm_flags & FIEMAP_FLAG_NOMERGE) {
+		uint64_t blksz = fm->fm_block_size;
+
+		for (uint64_t i = start; i < start + size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_DELALLOC | FIEMAP_EXTENT_UNKNOWN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, start, size, 0, 0, 0,
+		    FIEMAP_EXTENT_DELALLOC | FIEMAP_EXTENT_UNKNOWN |
+		    FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * Pending free extents set FIEMAP_EXTENT_UNWRITTEN since they will be a hole.
+ * FIEMAP_EXTENT_DELALLOC is set to indicate it has not yet been written.  The
+ * FIEMAP_EXTENT_UNKNOWN flag must be set when FIEMAP_EXTENT_DELALLOC is set.
+ * Free extents are only inserted in to the first extent tree.
+ */
+static void
+zfs_fiemap_add_free(void *arg, uint64_t start, uint64_t size)
+{
+	zfs_fiemap_t *fm = (zfs_fiemap_t *)arg;
+	avl_tree_t *t = &fm->fm_extent_trees[0];
+
+	zfs_fiemap_clear(t, start, size);
+
+	if (fm->fm_flags & FIEMAP_FLAG_NOMERGE) {
+		uint64_t blksz = fm->fm_block_size;
+
+		for (uint64_t i = start; i < start + size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+			    FIEMAP_EXTENT_UNKNOWN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, start, size, 0, 0, 0,
+		    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+		    FIEMAP_EXTENT_UNKNOWN | FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * The entire file is sparse and there are no level zero blocks with data.
+ * In this case pretend that hole block pointers exist to maintain consistency
+ * in the reported output.  Either add a single unwritten extent for the
```

Comment:
Current FIEMAP implementations return nothing for sparse files, which IMHO makes sense since there are no blocks allocated for the file...